core/src/main/java/org/frankframework/pipes/XmlSwitch.java

```diff
@@ -119,81 +116,82 @@ public void stop() {
+			return session.getString(getForwardNameSessionKey());
+		}
+
+		if (!(StringUtils.isEmpty(getXpathExpression()) && StringUtils.isEmpty(getStyleSheetName())) || StringUtils.isEmpty(getGetInputFromSessionKey())) {
 			try {
-				Map<String,Object> parametervalues = null;
+				Map<String, Object> parametervalues = null;
 				ParameterList parameterList = getParameterList();
+
 				if (!parameterList.isEmpty()) {
 					parametervalues = parameterList.getValues(message, session, isNamespaceAware()).getValueMap();
 				}
-				if(StringUtils.isNotEmpty(getSessionKey())) {
-					forward = transformerPool.transform(session.getMessage(getSessionKey()), parametervalues, isNamespaceAware());
-				} else {
-					message.preserve();
-					forward = transformerPool.transform(message, parametervalues, isNamespaceAware());
-				}
+
+				message.preserve();
+				return transformerPool.transform(message, parametervalues, isNamespaceAware());
 			} catch (Throwable e) {
 				throw new PipeRunException(this, "got exception on transformation", e);
 			}
-		} else if(StringUtils.isNotEmpty(getSessionKey())) {
-			forward = session.getString(getSessionKey());
 		}
-		log.debug("determined forward [{}]", forward);
 
-		if (StringUtils.isEmpty(forward) && getEmptyForwardName()!=null) {
+		if (StringUtils.isNotEmpty(getGetInputFromSessionKey())) {
+			try {
+				// Use the message as forward if none of the cases above apply
+				return message.asString();
+			} catch (IOException e) {
+				throw new PipeRunException(this, "Error reading message", e);
+			}
+		}
+
+		return null;
+	}
+
+	private PipeForward getPipeForward(String forward) {
+		PipeForward pipeForward;
+		if (StringUtils.isEmpty(forward) && getEmptyForwardName() != null) {
 			throwEvent(XML_SWITCH_FORWARD_FOUND_MONITOR_EVENT);
-			pipeForward=findForward(getEmptyForwardName());
+			pipeForward = findForward(getEmptyForwardName());
 		} else {
-
 			if (findForward(forward) != null) {
 				throwEvent(XML_SWITCH_FORWARD_FOUND_MONITOR_EVENT);
-				pipeForward=findForward(forward);
-			}
-			else {
+				pipeForward = findForward(forward);
+			} else {
 				log.info("determined forward [{}], which is not defined. Will use [{}] instead", forward, getNotFoundForwardName());
 				throwEvent(XML_SWITCH_FORWARD_NOT_FOUND_MONITOR_EVENT);
-				pipeForward=findForward(getNotFoundForwardName());
+				pipeForward = findForward(getNotFoundForwardName());
 			}
 		}
-
-		if (pipeForward==null) {
-			throw new PipeRunException (this, "cannot find forward or pipe named ["+forward+"]");
-		}
-		if(StringUtils.isNotEmpty(getStoreForwardInSessionKey())) {
-			session.put(getStoreForwardInSessionKey(), pipeForward.getName());
-		}
-
-		return new PipeRunResult(pipeForward, message);
-	}
-
-	@Override
-	public boolean consumesSessionVariable(String sessionKey) {
-		return super.consumesSessionVariable(sessionKey) || sessionKey.equals(getSessionKey());
+		return pipeForward;
```

Comment:
Of moet hier een leuke log regel neerzetten (forward [{}] resolved to [{}]), of redirect returnen 😄